src/org/exist/storage/XQueryPool.java

```diff
@@ -58,6 +56,8 @@
 
     private final static Logger LOG = LogManager.getLogger(XQueryPool.class);
 
+    private ConcurrentMap<Source, Deque<CompiledXQuery>> pool = new ConcurrentHashMap<>();
+
     private long lastTimeOutCheck;
     private long lastTimeOfCleanup;
 
```

Comment:
@shabanovd Okay you now have two `AtomicLong`s if they are used in an invariant together (i.e. in `returnObject`), this can give different results to having a single mutual exclusion around both variables. That might be acceptable, depending on the guarantees that you are offering on consistency. Have you considered this?